File: claycode/conversion/bits_tree/square.js
import { Tree } from "../../tree/tree.js"
import { TreeNode } from "../../tree/tree_node.js";

/* Util */

function bitStringToInt(bits) {
    bits.unshift("1"); // Add 1 at beginning to make sure it is an integer
    return Array.from(bits).reverse().reduce((acc, c, i) => acc + BigInt(c)*2n**BigInt(i), 0n);
}

function intToBitString(x) {
    console.assert(typeof x === "bigint");
    if (x <= 0n) {
        throw new Error("Input must be greater than 0");
    }
    return x.toString(2).slice(1); // Remove initial 1
}

function largestSquareBinsearch(x) {
    console.assert(typeof x === "bigint");
    if (x === 1n) {
        return 1n;
    }

    let l = 0n;
    let r = x;

    while (true) {
        const mid = l + ((r - l) / 2n);
        const midSq = mid * mid;

        if (midSq <= x && (mid + 1n) ** 2n > x) {
            return mid;
        } else {
            if (midSq <= x) {
                l = mid;
            } else {
                r = mid;
            }
        }
    }
}

function numberToSquareDecomposition(x) {
    console.assert(typeof x === "bigint");
    const decomposition = [];
    x -= 1n;

    while (x > 0n) {
        const lsq = largestSquareBinsearch(x);
        decomposition.push(lsq);
        x -= lsq ** 2n;
    }

    return decomposition;
}

function squareDecompositionToNumber(dec) {
    return dec.reduce((sum, num) => sum + num ** 2n, 1n); // Start accumulating from 1 to do +1
}

/* Encoding functions */

export function bitsToTree(bitsArray) {
    console.log("Encoding using squares");
    function populateTreeOfNumber(root, n) {
        const dec = numberToSquareDecomposition(n);
        root.children = dec.map(() => new TreeNode());
        for (let i = 0; i < root.children.length; i++) {
            populateTreeOfNumber(root.children[i], dec[i]);
        }
    }

    const root = new TreeNode();
    const n = bitStringToInt(bitsArray);
    populateTreeOfNumber(root, n);

    return new Tree(root);
}

export function treeToBits(tree) {
    function treeToNumber(root) {
        if (root.children.length === 0) {
            return 1;
        } else {
            const dec = root.children.map(c => treeToNumber(c));
            return squareDecompositionToNumber(dec);
        }
    }

    return intToBitString(treeToNumber(tree.root));
}

File: experiments/siggraph_2025_artifact/tree_lib/encodings/square.py

from tree_lib.util import bit_string_to_number, number_to_bit_string
from tree_lib.tree import TreeNode

######################
# Unordered tree encoding
# Decompose the number into (a^2 + b^2 + c^2) + 1 ...
######################

# Returns the largest square contained in `n`
def largest_square(n):
    if n == 1:
        return 1

    a = 0
    b = n
    while True:
        r = a + (b-a)//2
        rsq = r*r
        if rsq <= n and (r+1)**2 > n:
            return r
        else:
            if rsq <= n:
                a = r
            else:
                b = r

def number_to_square_decomposition(n):
    dec = []
    n-=1
    while n > 0:
        lsq = largest_square(n)
        dec += [lsq]
        n-=lsq**2
    return dec

def square_decomposition_to_number(dec):
    return sum([(num**2) for num in dec])+1

def bits_to_tree(bits: str) -> TreeNode: 
    def populate_tree_of_number(root: TreeNode, n):
        dec = number_to_square_decomposition(n)
        root.children = [TreeNode() for _ in range(len(dec))]
        for i,c in enumerate(root.children):
            populate_tree_of_number(c, dec[i])

    root = TreeNode()
    n = bit_string_to_number(bits)
    populate_tree_of_number(root, n)

    root.initialize()
    return root

def tree_to_bits(root: TreeNode) -> str:
    def tree_to_number(root: TreeNode):
        if root.children == []:
            return 1
        else:
            dec = [tree_to_number(c) for c in root.children]
            return square_decomposition_to_number(dec)
    
    return number_to_bit_string(tree_to_number(root))